### PR TITLE
perf: re-used prepared statements

### PR DESCRIPTION
**Pull request by jcosentino11**
*Issue #, if available:*

*Description of changes:*
Cache prepared statements, rather than creating new ones for every query. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java [hunk 1]
```diff
@@ -268,85 +371,73 @@ protected void setUpDatabase() throws SQLException {
...
+            return null;
+        }
     }
 
-    @SuppressWarnings("PMD.AvoidCatchingGenericException")
-    private <T> T performSqlOperation(CrashableFunction<Connection, T, SQLException> operation) throws SQLException {
-        try {
+    abstract class CachedStatement<T extends Statement, R> {
+        private T statement;
+        private Connection connection;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:395]: connect looks unused

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java [hunk 2]
```diff
@@ -82,10 +89,14 @@ public DiskSpoolDAO(NucleusPaths paths) throws IOException {
      * @throws SQLException if db is unable to be created
      */
     public void initialize() throws SQLException {
+        DriverManager.registerDriver(new org.sqlite.JDBC());
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:92]: do we need to register a new instance of a driver every time we make a new connection? Sounds unnecessary

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java (hunk 1 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:395]: Ah yeah, field is leftover from before using callbacks, will remove next time I'm at my laptop

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:92]: It's unnecessary but it also doesn't have an effect according to the docs, would need to see how much extra work it is or isn't doing

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:92]: Feels like a leak given you create a new driver each time.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java:92]: Yeah will look more closely, this is also existing behavior 🙃

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java (hunk 1 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java]: fixed

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java]: so turns out this line isn't even needed, driver will self-register anyway, removed it

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java]: > so turns out this line isn't even needed, driver will self-register anyway, removed it

this was wrong, not sure why it was working in local testing, but it failed UATs saying that driver's not found
```
java.io.IOException: java.sql.SQLException: No suitable driver found for jdbc:sqlite:/home/ec2-user/testResults/23a8fe84c2b77e4bcf9352b2925df0db9fd6d40b/work/aws.greengrass.DiskSpooler/spooler.db
```
Will add back in

**Pull request merged**